### ISSUE TITLE
Video upload

### ISSUE THREAD

**Issue by skiesx** [stanvanrooy/instauto#60]
Hi, @stanvanrooy ) 
Do you plan to implement video uploading?

**Comment by stanvanrooy** [comment]
Will be next up on the to-do list. That means it should be published to PyPI in 1-2 days.

**Comment by trashmit** [comment]
yes this would be amazing

**Comment by stanvanrooy** [comment]
Working on it :) If you watch this issue, you'll be notified as soon as it's published.

**Comment by stanvanrooy** [comment]
Sorry for the delay. I have a working version now, for videos of a small size. I'm still working on uploading longer/larger videos in phases.

**Comment by skiesx** [comment]
> 
> 
> Sorry for the delay. I have a working version now, for videos of a small size. I'm still working on uploading longer/larger videos in phases.

thx for this update
do you have an example code for posting a video?
and can you tell more about the small size? what are the limits...

**Comment by trashmit** [comment]
yeah ^ would be useful so I can implement it to some code preemptively instead of having to wait till you finish also lmk if you need any help

**Comment by stanvanrooy** [comment]
The method I had already implemented will be deprecated by ig, so I'll discard that work.

I've started work on segmented video uploads, should be ready soon.

Using it will be the same as the other methods.

**Comment by trashmit** [comment]
No problem, this might be useful to you https://github.com/dilame/instagram-private-api/blob/d77e228f61b717dd15225fc5247fd50846a49893/src/repositories/upload.repository.ts#L45

**Comment by stanvanrooy** [comment]
Possibly, but the flow for uploading a video in segments is pretty simple, but for some reason, ig doesn't accept the first request to the `/media/rupload_igvideo/{uuid4}?segmented=true&phase=start` endpoint. 

It always returns a 404. I'll push my code and the requests that need to be sent to a branch tonight.

**Comment by stanvanrooy** [comment]
See https://github.com/stanvanrooy/instauto/tree/60UploadVideos, especially the last commit.

These requests below are copies of the original ones. The current code fails on request 1. IG returns a 404.

**REQUEST 1:**
POST 			https://i.instagram.com/rupload_igvideo/{uuid}?segmented=true&phase=start HTTP/2.0
**HEADERS:**
:authority:		i.instagram.com
x-instagram-rupload-params:	{"upload_media_height":"480","xsharing_user_ids":"[\"4478472759\"]","upload_media_width":"480","upload_media_duration_ms":"18646","upload_id":"6664219672464","retry_context":"{\"num_reupload\":0,\"num_step_auto_retry\":0,\"num_step_manual_retry\":0}","media_type":"2"}
x-ig-connection-type:	WIFI
x-ig-capabilities:	3brTvw8=
x-ig-app-id:		567067343352427
user-agent:		Instagram 148.0.0.33.121 Android (29/10; 560dpi; 1440x2891; samsung; SM-G973F; beyond1; exynos9820; nl_NL; 227298996)
accept-language:	nl-NL, en-US
authorization:		Bearer IGT:2:==
x-mid:	****
ig-u-ig-direct-region-hint:	CLN
ig-u-shbid:		11577
ig-u-shbts:		1602745905.4249442
ig-u-ds-user-id:	2097017052
ig-u-rur:		PRN
accept-encoding:	zstd, gzip, deflate
x-fb-http-engine:	Liger
**RESPONSE:**
{"stream_id":3722315727800918,"status":"ok","xsharing_nonces": []}

**TWO REQUESTS FOR ALL SEGMENTS:**
GET 			https://i.instagram.com/rupload_igvideo/af821c235278fece7e5b888b13ecd45b-0-1998659?segmented=true&phase=transfer HTTP/2.0
**HEADERS:**
:authority:		i.instagram.com
stream-id:		3722315727800918
x-instagram-rupload-params:	{"upload_media_height":"480","xsharing_user_ids":"[]","upload_media_width":"480","upload_media_duration_ms":"18646","upload_id":"6664219672464","retry_context":"{\"num_reupload\":0,\"num_step_auto_retry\":0,\"num_step_manual_retry\":0}","media_type":"2"}
segment-start-offset:	0
x_fb_video_waterfall_id:	1602935971955
segment-type:		2  // 1 = audio, 2 = video, 3 = mixed, I believe
x-ig-connection-type:	WIFI
x-ig-capabilities:	3brTvw8=
x-ig-app-id:		567067343352427
user-agent:		Instagram 148.0.0.33.121 Android (29/10; 560dpi; 1440x2891; samsung; SM-G973F; beyond1; exynos9820; nl_NL; 227298996)
accept-language:	nl-NL, en-US
authorization:		Bearer IGT:2:==
x-mid:			****
ig-u-ig-direct-region-hint:	CLN
ig-u-shbid:		11577
ig-u-shbts:		1602745905.4249442
ig-u-ds-user-id:	2097017052
ig-u-rur:		PRN
accept-encoding:	zstd, gzip, deflate
x-fb-http-engine:	Liger
**RESPONSE:**
{"offset":0}

POST https://i.instagram.com/rupload_igvideo/{same uuid as GET request}?segmented=true&phase=transfer HTTP/2.0
**HEADERS:**
:authority:		i.instagram.com
x-entity-length:	1998659
x-entity-name:		af821c235278fece7e5b888b13ecd45b-0-1998659
stream-id:		3722315727800918
x-instagram-rupload-params:	{"upload_media_height":"480","xsharing_user_ids":"[]","upload_media_width":"480","upload_media_duration_ms":"18646","upload_id":"6664219672464","retry_context":"{\"num_reupload\":0,\"num_step_auto_retry\":0,\"num_step_manual_retry\":0}","media_type":"2"}
x-entity-type:		video/mp4
segment-start-offset:	0
x_fb_video_waterfall_id:	1602935971955
segment-type:		2
offset:			0
x-ig-connection-type:	WIFI
x-ig-capabilities:	3brTvw8=
x-ig-app-id:		567067343352427
user-agent:		Instagram 148.0.0.33.121 Android (29/10; 560dpi; 1440x2891; samsung; SM-G973F; beyond1; exynos9820; nl_NL; 227298996)
accept-language:	nl-NL, en-US
authorization:		Bearer IGT:2:==
x-mid:	                ****
ig-u-ig-direct-region-hint:	CLN
ig-u-shbid:		11577
ig-u-shbts:		1602745905.4249442
ig-u-ds-user-id:	2097017052
ig-u-rur:		        PRN
content-type:		application/octet-stream
content-length:		1998659
accept-encoding:	zstd, gzip, deflate
x-fb-http-engine:	Liger
**RESPONSE:**
{"status": "ok"}

**Comment by stanvanrooy** [comment]
Well, I found out what triggered the 404... the upload url is `https://i.instagram.com/rupload_igvideo/` and not `https://i.instagram.com/api/v1/rupload_igvideo/` :)

**Comment by kn0wm4d** [comment]
> Well, I found out what triggered the 404... the upload url is `https://i.instagram.com/rupload_igvideo/` and not `https://i.instagram.com/api/v1/rupload_igvideo/` :)

Awesome man, why are you starting a new codebase when you have [ping/instagram_private_api](https://github.com/ping/instagram_private_api/pull/307)? 

I did implement one of your methods to upload images there that has been deprecated, and it works. You could tweak that client and have tons of functionalities like post search and a lot more already done.

**Comment by trashmit** [comment]
Nice! Can't wait to see a working version

**Comment by trashmit** [comment]
any update?

**Comment by trashmit** [comment]
any update ?

**Comment by trashmit** [comment]
If you're too busy to do this just let us know if this feature will ever come

**Comment by skiesx** [comment]
@stanvanrooy hi, is there any good news? any update ?

**Comment by stanvanrooy** [comment]
I left my bag, with my laptop, in the bus, so all of my progress is pretty much gone. I'm currently really busy, so I don't have much time anymore to spend on this project.

Once it is finally implemented, I'll let you all know in this issue :) Thanks.

**Comment by stanvanrooy** [comment]
Little update on this. I've did a bit more work on it this evening (https://github.com/stanvanrooy/instauto/tree/60UploadVideos) and made quite a bit of progress.

Uploading the video is currently working, I just need to finish the configuring of the video. I'll try to spend a little bit more time on it this week.

**Comment by oitoito** [comment]
> Little update on this. I've did a bit more work on it this evening (https://github.com/stanvanrooy/instauto/tree/60UploadVideos) and made quite a bit of progress.
> 
> Uploading the video is currently working, I just need to finish the configuring of the video. I'll try to spend a little bit more time on it this week.

Are you going to merge this feature with main/master at some point? 

Thanks!
